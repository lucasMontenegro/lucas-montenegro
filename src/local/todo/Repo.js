const pgMinify = require("pg-minify")
const extractPgRowCount = require("local/utils/extractPgRowCount")
function minify (sql) {
  return pgMinify(sql, { compress: true, removeAll: true })
}
const templates = {
  create: minify(`
    INSERT INTO todo (user_id, description, priority)
    VALUES ($1, $2, $3)
  `),
  read: minify(`
    SELECT todo_id, description, priority, status, created_on, updated_on
    FROM todo
    WHERE user_id = $1
  `),
  update: minify(`
    UPDATE todo
    SET priority=$3, status=$4, updated_on=NOW()
    WHERE user_id = $1 AND todo_id = $2
  `),
  delete: minify(`
    DELETE FROM todo
    WHERE user_id = $1 AND todo_id = $2
  `),
}
module.exports = class TodoRepo {
  constructor(db, dc) {
    this.db = db
    this.pgp = db.$config.pgp
  }
  create (values) {
    return this.db.none(this.pgp.as.format(templates.create, values))
  }
  read (user_id) {
    return this.db.any(this.pgp.as.format(templates.read, user_id))
  }
  update (values) {
    return this.db.none(this.pgp.as.format(templates.update, values))
  }
  delete (values) {
    return this.db.result(this.pgp.as.format(templates.delete, values), null, extractPgRowCount)
  }
}