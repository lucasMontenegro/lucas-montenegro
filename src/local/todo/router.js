const express = require("express")
const jwt = require("express-jwt")
const jwksRsa = require("jwks-rsa")
const jwtAuthz = require("express-jwt-authz")
const bodyParser = require("body-parser")
const requestHandlers = require("local/todo/requestHandlers")
const validators = require("local/todo/validators")
const { process } = global
const todo = module.exports = express.Router()
todo.use(bodyParser.json())
todo.use(bodyParser.urlencoded({ extended: true }))
todo.route(`/todo`)
  .all(jwt({
    secret: jwksRsa.expressJwtSecret({
      cache: true,
      rateLimit: true,
      jwksRequestsPerMinute: 5,
      jwksUri: `${process.env.AUTH0_DOMAIN}/.well-known/jwks.json`,
    }),
    audience: `${process.env.LUCAS_MONTENEGRO_DOMAIN}/api/todo`,
    issuer: process.env.LUCAS_MONTENEGRO_DOMAIN,
    algorithms: [`RS256`],
  }))
  .post(jwtAuthz([`create:todo`]), validators.post, requestHandlers.post)
  .get(jwtAuthz([`read:todo`]), requestHandlers.get)
  .put(jwtAuthz([`update:todo`]), validators.put, requestHandlers.put)
  .delete(jwtAuthz([`delete:todo`]), validators.delete, requestHandlers.delete)