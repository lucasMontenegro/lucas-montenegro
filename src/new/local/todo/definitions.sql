--
-- PostgreSQL database dump
--

-- Dumped from database version 11.7 (Debian 11.7-2.pgdg90+1)
-- Dumped by pg_dump version 11.7 (Debian 11.7-2.pgdg90+1)

SET statement_timeout = 0;
SET lock_timeout = 0;
SET idle_in_transaction_session_timeout = 0;
SET client_encoding = 'UTF8';
SET standard_conforming_strings = on;
SELECT pg_catalog.set_config('search_path', '', false);
SET check_function_bodies = false;
SET xmloption = content;
SET client_min_messages = warning;
SET row_security = off;

--
-- Name: public; Type: SCHEMA; Schema: -; Owner: -
--

CREATE SCHEMA "public";


--
-- Name: SCHEMA "public"; Type: COMMENT; Schema: -; Owner: -
--

COMMENT ON SCHEMA "public" IS 'standard public schema';


--
-- Name: todo_priority; Type: TYPE; Schema: public; Owner: -
--

CREATE TYPE "public"."todo_priority" AS ENUM (
    'high',
    'medium',
    'low'
);


--
-- Name: todo_status; Type: TYPE; Schema: public; Owner: -
--

CREATE TYPE "public"."todo_status" AS ENUM (
    'done',
    'pending',
    'cancelled'
);


SET default_tablespace = '';

SET default_with_oids = false;

--
-- Name: todo; Type: TABLE; Schema: public; Owner: -
--

CREATE TABLE "public"."todo" (
    "todo_id" bigint NOT NULL,
    "user_id" "text" NOT NULL,
    "description" character varying(255) NOT NULL,
    "priority" "public"."todo_priority" DEFAULT 'medium'::"public"."todo_priority" NOT NULL,
    "status" "public"."todo_status" DEFAULT 'pending'::"public"."todo_status" NOT NULL,
    "updated_on" timestamp with time zone,
    "created_on" timestamp with time zone DEFAULT CURRENT_TIMESTAMP
);


--
-- Name: todo_todo_id_seq; Type: SEQUENCE; Schema: public; Owner: -
--

CREATE SEQUENCE "public"."todo_todo_id_seq"
    START WITH 1
    INCREMENT BY 1
    NO MINVALUE
    NO MAXVALUE
    CACHE 1;


--
-- Name: todo_todo_id_seq; Type: SEQUENCE OWNED BY; Schema: public; Owner: -
--

ALTER SEQUENCE "public"."todo_todo_id_seq" OWNED BY "public"."todo"."todo_id";


--
-- Name: todo todo_id; Type: DEFAULT; Schema: public; Owner: -
--

ALTER TABLE ONLY "public"."todo" ALTER COLUMN "todo_id" SET DEFAULT "nextval"('"public"."todo_todo_id_seq"'::"regclass");


--
-- Name: todo todo_pkey; Type: CONSTRAINT; Schema: public; Owner: -
--

ALTER TABLE ONLY "public"."todo"
    ADD CONSTRAINT "todo_pkey" PRIMARY KEY ("todo_id");


--
-- Name: user_id__idx; Type: INDEX; Schema: public; Owner: -
--

CREATE INDEX "user_id__idx" ON "public"."todo" USING "btree" ("user_id");


--
-- PostgreSQL database dump complete
--

