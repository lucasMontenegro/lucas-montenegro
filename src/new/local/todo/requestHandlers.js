import minifySql from "new/local/utils/minifySql"
import syncRequestHandler from "new/local/utils/syncRequestHandler"
import { db, pgp } from "new/local/postgres"
import extractRowCount from "new/local/utils/postgres/extractRowCount"
const templates = {
  create: minifySql(`
    INSERT INTO todo (user_id, description, priority)
    VALUES ($1, $2, $3)
  `),
  read: minifySql(`
    SELECT todo_id, description, priority, status, created_on, updated_on
    FROM todo
    WHERE user_id = $1
  `),
  update: minifySql(`
    UPDATE todo
    SET priority=$3, status=$4, updated_on=NOW()
    WHERE user_id = $1 AND todo_id = $2
  `),
  delete: minifySql(`
    DELETE FROM todo
    WHERE user_id = $1 AND todo_id = $2
  `),
}
export default {
  post: syncRequestHandler(async (req, res) => {
    const { body } = req
    await db.none(pgp.as.format(templates.create, [req.user.sub, body.description, body.priority]))
    res.end()
  }),
  get: syncRequestHandler(async (req, res) => {
    res.json(db.any(pgp.as.format(templates.read, req.user.sub)))
  }),
  put: syncRequestHandler(async (req, res) => {
    const { body } = req
    const values = [req.user.sub, req.params.todo_id, body.priority, body.status]
    await db.none(pgp.as.format(templates.update, values))
    res.end()
  }),
  delete: syncRequestHandler(async (req, res) => {
    const { body } = req
    const values = [req.user.sub, req.params.todo_id]
    await db.result(pgp.as.format(templates.delete, values), null, extractRowCount)
    res.end()
  }),
}