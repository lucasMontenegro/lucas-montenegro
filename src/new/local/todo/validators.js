import { body, param } from "express-validator"
const description = body(`description`)
  .isString()
  .withMessage(`Invalid to-do item description`)
  .isLength({ min: 1 })
  .withMessage(`The to-do item description must be at least one character long`)
  .isLength({ max: 256 })
  .withMessage(`The to-do item description must be at most 256 characters long`)
const priority = body(`priority`)
  .isIn([`high`, `medium`, `low`])
  .withMessage(`The to-do item priority must be "high", "medium" or "low"`)
const todo_id = param(`todo_id`)
  .isNumeric({ no_symbols: true })
  .withMessage(`The to-do item ID must only contain numbers`)
const status = body(`status`)
  .isIn([`done`, `pending`])
  .withMessage(`The to-do item status must be "done" or "pending"`)
export default {
  post: [description, priority],
  put: [todo_id, priority, status],
  delete: todo_id,
}