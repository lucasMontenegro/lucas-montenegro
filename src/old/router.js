import React from "react"
import { Redirect } from "react-router-dom"
import l10n from "./l10n"
import LanguageNotFound from "./LanguageNotFound"
import Nav from "./Nav"
import Frame from "./Frame"

export default routerProps => {
  const { location, history } = routerProps
  const { pathname } = location
  if (pathname === '/') {
    return <Redirect to={l10n.english.home} />
  }
  const cap = /^\/([^/]+)/.exec(pathname)
  const lang = cap && cap[1]
  const langRoute = lang && l10n[lang]
  if (!langRoute) {
    return <LanguageNotFound history={history} />
  }
  if (pathname.length <= lang.length + 2) {
    return <Redirect to={langRoute.home} />
  }
  const innerRoute = Object.values(langRoute.routes).find(route => route.match(location))
  const nav = <Nav render={langRoute.nav} history={history} />
  if (!innerRoute) {
    return <Frame render={langRoute.notFound} nav={nav} />
  }
  return <Frame render={innerRoute.render} nav={nav} />
}
