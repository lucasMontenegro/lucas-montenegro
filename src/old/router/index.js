import React, { Fragment } from "react"
import { Redirect } from "react-router-dom"
import routes from "../routes"

const persistent = routes.filter(route => route.persistent)
export default props => {
  const main = routes.filter(route => route.match(props.location))
  return (
    <Fragment>
      <main.Component show key={main.key} {...props} />
      {persistent.map(route => <route.Component key={route.key} {...props} />)}
    </Fragment>
  )
}
