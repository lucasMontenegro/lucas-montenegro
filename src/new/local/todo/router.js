import express from "express"
import jwt from "express-jwt"
import jwksRsa from "jwks-rsa"
import jwtAuthz from "express-jwt-authz"
import bodyParser from "body-parser"
import requestHandlers from "new/local/todo/requestHandlers"
import validators from "new/local/todo/validators"
import globals from "new/local/utils/globals"
const todo = express.Router()
todo.use(bodyParser.json())
todo.use(bodyParser.urlencoded({ extended: true }))
todo.route(`/todo`)
  .all(jwt({
    secret: jwksRsa.expressJwtSecret({
      cache: true,
      rateLimit: true,
      jwksRequestsPerMinute: 5,
      jwksUri: `${globals.process.env.AUTH0_DOMAIN}/.well-known/jwks.json`,
    }),
    audience: `${globals.process.env.LUCAS_MONTENEGRO_DOMAIN}/api/todo`,
    issuer: globals.process.env.LUCAS_MONTENEGRO_DOMAIN,
    algorithms: [`RS256`],
  }))
  .post(jwtAuthz([`create:todo`]), validators.post, requestHandlers.post)
  .get(jwtAuthz([`read:todo`]), requestHandlers.get)
  .put(jwtAuthz([`update:todo`]), validators.put, requestHandlers.put)
  .delete(jwtAuthz([`delete:todo`]), validators.delete, requestHandlers.delete)
export default todo