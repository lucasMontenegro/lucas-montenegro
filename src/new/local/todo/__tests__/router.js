jest.mock(`new/local/utils/globals`, () => ({
  __esModule: true,
  default: {
    process: {
      env: {
        AUTH0_DOMAIN: `https://lucas-montenegro.auth0.com`,
        LUCAS_MONTENEGRO_DOMAIN: `https://lucasmontenegro.com.ar`,
      },
    },
  },
}))
jest.mock(`new/local/todo/validators`, () => ({
  __esModule: true,
  default: [`post`, `put`, `delete`].reduce((obj, name) => {
    obj[name] = () => {}
    return obj
  }, {}),
}))
jest.mock(`new/local/todo/requestHandlers`, () => ({
  __esModule: true,
  default: [`post`, `get`, `put`, `delete`].reduce((obj, name) => {
    obj[name] = () => {}
    return obj
  }, {}),
}))
describe(`new/local/todo/router`, () => {
  it(`should run`, () => {
    require("new/local/todo/router")
  })
})