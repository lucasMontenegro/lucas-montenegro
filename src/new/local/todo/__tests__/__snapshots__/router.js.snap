// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`new/local/todo/router should utilize the dependency APIs correctly 1`] = `
Object {
  "body-parser": "^1.19.0",
  "express": "^4.17.1",
  "express-jwt": "^5.3.1",
  "express-jwt-authz": "^2.3.1",
  "jwks-rsa": "^1.6.0",
}
`;
